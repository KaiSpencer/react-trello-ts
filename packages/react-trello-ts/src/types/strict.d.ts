declare module "uuid/v1" {
	export default function uuid(): string;
}
