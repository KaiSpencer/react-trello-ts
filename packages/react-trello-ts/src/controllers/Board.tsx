import React, { createContext, FC, PropsWithChildren } from "react";
import classNames from "classnames";
import uuidv1 from "uuid/v1";
import { BoardContainer } from "./BoardContainer";
import * as DefaultComponets from "../components";
import { BoardData } from "../types/Board";
import { store } from "../store/store";

export const Board: FC<
	PropsWithChildren<{
		id?: string;
		className?: string;
		components: typeof DefaultComponets;
		data: BoardData;
		t?: any;
	}>
> = ({ data, children, className, components, id, t, ...rest }) => {
	const allClassNames = classNames("react-trello-board", className || "");

	return (
		<BoardContext.Provider value={store}>
			<components.GlobalStyle />
			<BoardContainer
				id={id || uuidv1()}
				data={data}
				components={components}
				t={t}
				className={allClassNames}
				{...rest}
			/>
		</BoardContext.Provider>
	);
};

export const BoardContext = createContext(store);
