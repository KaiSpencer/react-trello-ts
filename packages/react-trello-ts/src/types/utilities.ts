/**
 * Given a type `TType` and a type `TRequired` which is a subset of required keys of `TType`, return a new type which is the
 * same as `TType` but with all keys optional except those in `TRequired`.
 */
export type PartialExcept<
	TType,
	TRequired extends keyof TType,
> = Partial<TType> & Pick<TType, TRequired>;
