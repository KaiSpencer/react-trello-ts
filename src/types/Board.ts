import {CSSProperties} from 'react'

export interface BoardData {
  lanes: Lane[]
}

export interface Lane {
  id: string
  title?: string
  label?: string
  style?: CSSProperties
  cards: Card[]
  currentPage?: number
  droppable?: boolean
  labelStyle?: CSSProperties
  cardStyle?: CSSProperties
  disallowAddingCard?: boolean
  [key: string]: any
}

export interface Card {
  id: string
  title?: string
  label?: string
  description?: string
  laneId?: string
  style?: CSSProperties
  draggable?: boolean
  [key: string]: any
}
