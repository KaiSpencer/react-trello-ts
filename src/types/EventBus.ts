import {MapLaneHelperToReducer} from 'rt/reducers/BoardReducer'

export interface EventBusHandle {
  publish: (event: EventBusEvent) => any
}

type EventBusEvent = {
  type: keyof typeof MapLaneHelperToReducer
}
