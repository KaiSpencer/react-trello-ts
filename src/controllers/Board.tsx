import classNames from "classnames";
import React, { FC, PropsWithChildren, createContext } from "react";
import { v1 } from "uuid";
import * as DefaultComponets from "../components";
import { store } from "../store/store";
import { BoardData } from "../types/Board";
import { BoardContainer } from "./BoardContainer";

export const Board: FC<
	PropsWithChildren<{
		id?: string;
		className?: string;
		components: typeof DefaultComponets;
		data: BoardData;
		t?: any;
	}>
> = ({ data, children, className, components, id, t, ...rest }) => {
	const allClassNames = classNames("react-trello-board", className || "");

	return (
		<BoardContext.Provider value={store}>
			{/*
			// @ts-ignore */}
			<components.GlobalStyle />
			<BoardContainer
				id={id || v1()}
				data={data}
				components={components}
				t={t}
				className={allClassNames}
				{...rest}
			/>
		</BoardContext.Provider>
	);
};

export const BoardContext = createContext(store);
