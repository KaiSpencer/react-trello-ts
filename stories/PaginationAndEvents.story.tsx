import { Meta, StoryObj, storiesOf } from "@storybook/react";
import React, { FC, PropsWithChildren } from "react";

import Board from "../src";
import { BoardData } from "../src/types/Board";
import { EventBusHandle } from "../src/types/EventBus";

const PER_PAGE = 15;
Board.displayName = "Board";

function generateCards(requestedPage = 1) {
	const cards = [];
	const fetchedItems = (requestedPage - 1) * PER_PAGE;
	for (let i = fetchedItems + 1; i <= fetchedItems + PER_PAGE; i++) {
		cards.push({
			id: `${i}`,
			title: `Card${i}`,
			description: `Description for #${i}`,
			metadata: { cardId: `${i}` },
		});
	}
	return cards;
}
const delayedPromise = (durationInMs, resolutionPayload) => {
	return new Promise((resolve) => {
		setTimeout(() => {
			resolve(resolutionPayload);
		}, durationInMs);
	});
};
const BoardWrapper: FC<PropsWithChildren<{ data: BoardData }>> = ({
	data,
	children,
}) => {
	const [boardData, setBoardData] = React.useState(data);
	const [eventBus, setEventBus] = React.useState<EventBusHandle>();
	const onDataChange = (newData: BoardData) => {
		setBoardData(newData);
	};
	const refreshCards = () => {
		setBoardData({
			lanes: [
				{
					id: "Lane1",
					title: "Changed Lane",
					cards: [],
				},
			],
		});
	};
	const addCard = () => {
		eventBus.publish({
			type: "ADD_CARD",
			laneId: "Lane1",
			card: {
				id: "000",
				title: "EC2 Instance Down",
				label: "30 mins",
				description: "Main EC2 instance down",
				metadata: { cardId: "000" },
			},
		});
	};
	const paginate = (requestedPage, laneId) => {
		const newCards = generateCards(requestedPage);
		return delayedPromise(2000, newCards);
	};

	return (
		<div>
			<button onClick={addCard} style={{ margin: 5 }}>
				Add Card
			</button>
			<button onClick={refreshCards} style={{ margin: 5 }}>
				Refresh Board
			</button>
			<Board
				data={boardData}
				eventBusHandle={(eventBus) => setEventBus(eventBus)}
				laneSortFunction={(card1, card2) =>
					parseInt(card1.id) - parseInt(card2.id)
				}
				onLaneScroll={paginate}
				onDataChange={onDataChange}
			/>
		</div>
	);
};

export default ({
	title: "Advanced Features",
	component: Board,
} satisfies Meta<typeof Board>);

type Story = StoryObj<typeof Board>;

export const ScrollingAndEvents: Story = {
	name: "Scrolling and Events",
	render: (args) => {
		const data = {
			lanes: [
				{
					id: "Lane1",
					title: "Lane1",
					cards: generateCards(),
				},
			],
		};

		return <BoardWrapper data={data} />;
	},
};
