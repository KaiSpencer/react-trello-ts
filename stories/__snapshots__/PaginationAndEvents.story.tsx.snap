// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced Features ScrollingAndEvents smoke-test 1`] = `
<div>
  <button style="margin: 5px;">
    Add Card
  </button>
  <button style="margin: 5px;">
    Refresh Board
  </button>
  <div draggable="false"
       class="sc-iGgWBj iSNfA"
  >
    <div>
      <div class="smooth-dnd-container horizontal">
        <section draggable="false"
                 class="sc-kAyceB dgWEjS react-trello-lane"
        >
          <header class="sc-gsFSXq sc-imWYAI imRIDz gLxuAH">
            <span class="sc-kpDqfm gDkBAH">
              Lane1
            </span>
          </header>
          <div class="sc-dhKdcB ikKiDc">
            <div class="smooth-dnd-container vertical">
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="1"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card1
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #1
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="2"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card2
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #2
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="3"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card3
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #3
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="4"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card4
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #4
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="5"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card5
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #5
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="6"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card6
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #6
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="7"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card7
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #7
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="8"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card8
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #8
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="9"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card9
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #9
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="10"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card10
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #10
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="11"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card11
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #11
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="12"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card12
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #12
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="13"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card13
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #13
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="14"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card14
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #14
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="15"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                >
                  <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                    <span draggable="true"
                          class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                    >
                      Card15
                    </span>
                    <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                    </span>
                    <div class="sc-hmdomO gJuJJJ">
                      <button class="sc-jsJBEP djrAfI">
                        ✖
                      </button>
                    </div>
                  </header>
                  <div class="sc-eDPEul cRhiim">
                    Description for #15
                  </div>
                </article>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
</div>
`;
